### PR TITLE
feat(SGC-101): query flamingo db to select WA API Client

### PR DESCRIPTION
**Pull request by tan-zx**
[Ticket](https://linear.app/ogp/issue/SGC-101/query-flamingo-db-to-determine-wa-api-client-to-use) and [explanation on decision to not use Sequelize here](https://www.notion.so/opengov/RFC-for-WhatsApp-Integration-c04a169adc9d452b9e37aaef8bbd8e72?pvs=4#b60c798c355444dca5c35ec794f38582)

- [x] Test on local
- [x] Test on staging

Deployment notes:
- [x] Make necessary infra changes on Flamingo DB to allow connection (both staging and production)
  - NB for local dev, connect with OGP VPN turned one
- [ ] Add env var `FLAMINGO_DB_URI`

### CODE REVIEW

File: shared/tsconfig.json [hunk 1]
```diff
@@ -31,6 +31,7 @@
     "esModuleInterop": true,
     "experimentalDecorators": true,
     "emitDecoratorMetadata": true,
-    "composite": true
+    "composite": true,
+    "skipLibCheck": true
```
Comment (tan-zx) [inline shared/tsconfig.json:35]: needed to solve a TS error with Drizzle ORM (suspect it's to do with diverging TS version?)

File: shared/src/clients/flamingo-db-client.class/index.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  private dbUri: string
+  constructor(dbUri: string) {
+    this.dbUri = dbUri
+  }
+
+  private async initializeDb(): Promise<NodePgDatabase> {
+    const client = new Client({
+      connectionString: this.dbUri,
+      ssl: {
+        rejectUnauthorized: false,
```
Comment (tan-zx) [inline shared/src/clients/flamingo-db-client.class/index.ts:16]: necessary setting to connect to Flamingo DB

File: shared/src/clients/flamingo-db-client.class/index.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      },
+    })
+    await client.connect()
+    return drizzle(client)
+  }
+  public async getAssociatedApiClientId(
+    phoneNumber: string
+  ): Promise<number | null> {
+    const db = await this.initializeDb()
```
Comment (tan-zx) [inline shared/src/clients/flamingo-db-client.class/index.ts:26]: initializing db here to get around the fact that you cannot have an async constructor.

I'm pretty sure it's not a good idea to initialize and connect to the db on each method call (?) but I'm also not sure what's a better way to deal with this. Stopgap solution, should be fine at a small scale, should remove afterwards?

Comment (tan-zx) [reply shared/src/clients/flamingo-db-client.class/index.ts]: fixed I think

File: backend/src/govsg/services/govsg-transactional.service.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    components: params,
+    apiClient,
+    language: WhatsAppLanguages.english,
+  }
+  // differential treatment based on local vs staging/prod
+  // because WA API Client is inaccessible from local
+  const isLocal = config.get('env') === 'development'
+  await WhatsAppService.whatsappClient
+    .validateSingleRecipient(messageToSend, isLocal)
+    .catch((err) => {
```
Comment (stanleynguyen) [inline backend/src/govsg/services/govsg-transactional.service.ts:51]: can we use try/catch here? I have a strong opinion of not mixing Promises and async/await

File: backend/src/govsg/services/govsg-transactional.service.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        message: 'Error validating recipient',
+        action,
+        error: err,
+      })
+      throw new ApiInvalidRecipientError(err.message)
+    })
+
+  const messageId = await WhatsAppService.whatsappClient
+    .sendMessage(messageToSend, isLocal)
+    .catch((err) => {
```
Comment (stanleynguyen) [inline backend/src/govsg/services/govsg-transactional.service.ts:62]: similar point to above about try/catch vs Promises

File: backend/src/govsg/services/govsg-transactional.service.ts (hunk 4 above)
Comment (tan-zx) [reply backend/src/govsg/services/govsg-transactional.service.ts:51]: > I have a strong opinion of not mixing Promises and async/await

why/wdym? I think this helps to reduce the indentation and makes it clearer where the error occurs tho

**Comment by tan-zx** [comment]
closing this PR — to reopen a separate PR that doesn't use Drizzle ORM. keep things light and simple first

**Pull request closed without merging**